File: 14. Tree/Binary_Tree_Inorder_Traversal.java
// Binary Tree Inorder Traversal
// Iterative

class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        if(root == null)
            return list;
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode temp = root;
        do{
            while(temp!=null){
                stack.push(temp);
                temp=temp.left;
            }
            temp = stack.pop();
            
            while(temp!=null){
                list.add(temp.val);
                if(temp.right!=null){
                    temp=temp.right;
                    break;
                }
                if(!stack.isEmpty())
                    temp=stack.pop();
                else
                    return list;
            }
        }while(temp!=null);
        return list;
    }
}

File: 27. LeetCode Problems/binary-tree-inorder-traversal.cpp
class Solution {
public:
vector inorderTraversal(TreeNode* root) {
vector result;
TreeNode *curr = root;
while(curr != NULL){
if(curr->left == NULL){
result.push_back(curr->val);
curr = curr->right;
}
else{
TreeNode * pred = curr ->left;
while(pred->right != curr && pred->right != NULL)
pred = pred->right;

            if(pred->right == NULL){
                pred->right = curr;
                curr = curr->left;
            }
            else{
                pred -> right = NULL;
                result.push_back(curr->val);
                curr = curr->right;
            }
        }
    }
     return result;   
}
};